File: src/cpu/unofficial.rs
//! Unofficial / illegal 6502 opcodes (M33) — dispatch root.
//!
//! The 6502 has 151 official opcodes and 105 unofficial ("illegal") opcodes
//! that are the result of PLA decode don't-cares. Many commercial and
//! homebrew NES titles rely on them for compact code, so a complete
//! emulator must implement them.
//!
//! This module is the dispatch root: [`Cpu::execute_unofficial`] handles
//! the "simple" unofficial opcodes (NOP variants, LAX, SAX, the immediate
//! combined ops ANC/ALR/ARR/AXS/XAA, LAS, and KIL) and delegates the
//! RMW-combo opcodes (DCP/ISC/SLO/RLA/SRE/RRA) and the unstable indexed
//! stores (TAS/AHX/SHX/SHY) to [`super::unofficial_rmw::Cpu::execute_unofficial_rmw`].
//! The combo/imm/store *helpers* live in [`super::unofficial_special`].
//!
//! # Cycle counts
//!
//! Cycle counts follow the NESdev reference table at
//! <https://www.nesdev.org/wiki/CPU_unofficial_opcodes>. RMW-combo opcodes
//! use the RMW addressing helpers and so inherit the unconditional
//! dummy-read cycle (no page-cross penalty on the indexed variants — the
//! +1 is already paid). NOP absolute,X is the exception: it is a
//! *read*-style NOP and adds the +1 page-cross penalty.
//!
//! # Unstable opcodes
//!
//! XAA, LAS, TAS, AHX, SHX, and SHY are documented as "highly unstable" on
//! real silicon — their exact behavior varies by chip revision,
//! temperature, and prior bus state. The implementations here follow the
//! most commonly emulated model (per NESdev wiki) so that the majority of
//! test ROMs and commercial titles behave correctly. Determinism is
//! preserved: the same inputs always produce the same outputs on this
//! emulator.
//!
//! See: https://www.nesdev.org/wiki/CPU_unofficial_opcodes
//! See: https://www.nesdev.org/6502.txt — "Unofficial opcodes" section.

use super::{Cpu, Dummy};
use crate::bus::Bus;

impl Cpu {
    /// Decode and execute an unofficial opcode. Called from the fallback
    /// arm of [`super::opcodes::Cpu::execute`] for any opcode byte that does
    /// not match an official instruction. Returns the cycle count. Any byte
    /// that matches neither an official nor an unofficial opcode (there are
    /// none in the 0x00..=0xFF space once both sets are covered, but the
    /// fallback is kept defensive) returns a 2-cycle NOP so the hot loop
    /// never panics.
    ///
    /// See: https://www.nesdev.org/wiki/CPU_unofficial_opcodes
    pub fn execute_unofficial(&mut self, bus: &mut Bus, opcode: u8) -> u8 {
        match opcode {
            // ---- NOP variants ------------------------------------------
            // Implied NOPs (2 cycles): 0x1A, 0x3A, 0x5A, 0x7A, 0xDA, 0xFA.
            0x1A | 0x3A | 0x5A | 0x7A | 0xDA | 0xFA => 2,
            // Immediate NOPs (2 cycles): 0x80, 0x82, 0x89, 0xC2, 0xE2.
            0x80 | 0x82 | 0x89 | 0xC2 | 0xE2 => {
                let _ = self.fetch_byte(bus);
                2
            }
            // Zero-page NOPs (3 cycles): 0x04, 0x44, 0x64.
            0x04 | 0x44 | 0x64 => {
                let _ = self.fetch_byte(bus);
                3
            }
            // Zero-page,X NOPs (4 cycles): 0x14, 0x34, 0x54, 0x74, 0xD4, 0xF4.
            0x14 | 0x34 | 0x54 | 0x74 | 0xD4 | 0xF4 => {
                let _ = self.am_zero_page_x(bus, Dummy::Rmw);
                4
            }
            // Absolute NOP (4 cycles): 0x0C.
            0x0C => {
                let _ = self.fetch_word(bus);
                4
            }
            // Absolute,X NOPs (4 + page-cross cycles): 0x1C, 0x3C, 0x5C,
            // 0x7C, 0xDC, 0xFC. These are *read*-style NOPs: the dummy read
            // fires on page cross and the +1 penalty applies.
            0x1C | 0x3C | 0x5C | 0x7C | 0xDC | 0xFC => {
                let (_, pc) = self.am_absolute_x(bus, Dummy::Read);
                4 + u8::from(pc)
            }

            // ---- LAX (load A and X) ------------------------------------
            0xA7 => {
                let v = self.rd_zp(bus);
                self.lax(v);
                3
            }
            0xB7 => {
                let v = self.rd_zp_y(bus);
                self.lax(v);
                4
            }
            0xAF => {
                let v = self.rd_abs(bus);
                self.lax(v);
                4
            }
            0xBF => {
                let (v, pc) = self.rd_abs_y(bus);
                self.lax(v);
                4 + u8::from(pc)
            }
            0xA3 => {
                let v = self.rd_ind_x(bus);
                self.lax(v);
                6
            }
            0xB3 => {
                let (v, pc) = self.rd_ind_y(bus);
                self.lax(v);
                5 + u8::from(pc)
            }

            // ---- SAX (store A & X) -------------------------------------
            0x87 => {
                self.wr_zp(bus, self.a & self.x);
                3
            }
            0x97 => {
                self.wr_zp_y(bus, self.a & self.x);
                4
            }
            0x8F => {
                self.wr_abs(bus, self.a & self.x);
                4
            }
            0x83 => {
                self.wr_ind_x(bus, self.a & self.x);
                6
            }

            // ---- ANC (AND then copy N to C) ----------------------------
            // 0x0B and 0x2B are both ANC. A = A & imm; C = N = bit 7 of A.
            0x0B | 0x2B => {
                let v = self.rd_imm(bus);
                self.anc(v);
                2
            }

            // ---- ALR (AND then LSR) ------------------------------------
            // A = (A & imm) >> 1; C = old bit 0 of (A & imm).
            0x4B => {
                let v = self.rd_imm(bus);
                self.alr(v);
                2
            }

            // ---- ARR (AND then ROR, special V/C) -----------------------
            0x6B => {
                let v = self.rd_imm(bus);
                self.arr(v);
                2
            }

            // ---- AXS / SBX (subtract imm from A&X, store in X) ---------
            0xCB => {
                let v = self.rd_imm(bus);
                self.axs(v);
                2
            }

            // ---- XAA (unstable: A = (A | magic) & X & imm) -------------
            0x8B => {
                let v = self.rd_imm(bus);
                self.xaa(v);
                2
            }

            // ---- LAS / LAR (A = X = SP = M & SP) -----------------------
            0xBB => {
                let (v, pc) = self.rd_abs_y(bus);
                let r = v & self.sp;
                self.a = r;
                self.x = r;
                self.sp = r;
                self.set_nz(r);
                4 + u8::from(pc)
            }

            // ---- KIL / JAM / HLT (halt the CPU) ------------------------
            // 0x02, 0x12, 0x22, 0x32, 0x42, 0x52, 0x62, 0x72, 0x92, 0xB2,
            // 0xD2, 0xF2. The CPU freezes; only RESET revives it. We
            // consume 1 cycle and set the halted flag — `step` becomes a
            // 1-cycle no-op thereafter so the PPU/APU keep advancing.
            0x02 | 0x12 | 0x22 | 0x32 | 0x42 | 0x52 | 0x62 | 0x72 | 0x92 | 0xB2 | 0xD2 | 0xF2 => {
                self.set_halted(true);
                1
            }

            // ---- RMW-combo + unstable stores ---------------------------
            // DCP, ISC, SLO, RLA, SRE, RRA, TAS, AHX, SHX, SHY are handled
            // by the sibling dispatch in `unofficial_rmw.rs`.
            _ => self.execute_unofficial_rmw(bus, opcode),
        }
    }

    /// LAX — load A and X simultaneously from `m`. Sets N/Z from `m`.
    fn lax(&mut self, m: u8) {
        self.a = m;
        self.x = m;
        self.set_nz(m);
    }
}

File: cores/go/internal/cpu/unofficial.go
// unofficial.go - unofficial / illegal 6502 opcodes (port of cpu_unofficial.c).
package cpu

import "nes-core-go/internal/bus"

const xaaMagic uint8 = 0x00

// rmwCombo mirrors the rmw wrapper (read, transform, write, set N/Z).
func (c *Cpu) rmwCombo(b *bus.Bus, op Operand, f valueFn) {
	v := c.readOperand(b, op)
	neu := f(c, v)
	c.writeOperand(b, op, neu)
	c.setNz(neu)
}

// ---- Immediate combined ops ----

func (c *Cpu) anc(m uint8) {
	c.A &= m
	c.setNz(c.A)
	c.SetCarry((c.A & 0x80) != 0)
}

func (c *Cpu) alr(m uint8) {
	v := c.A & m
	c.SetCarry((v & 0x01) != 0)
	result := v >> 1
	c.A = result
	c.setNz(result)
}

func (c *Cpu) arr(m uint8) {
	v := c.A & m
	var carryIn uint8
	if c.Carry() {
		carryIn = 0x80
	}
	result := (v >> 1) | carryIn
	c.A = result
	c.setNz(result)
	c.SetCarry((result & 0x40) != 0)
	c.SetOverflow(((result ^ (result << 1)) & 0x40) != 0)
}

func (c *Cpu) axs(m uint8) {
	ax := c.A & c.X
	c.SetCarry(ax >= m)
	result := ax - m
	c.X = result
	c.setNz(result)
}

func (c *Cpu) xaa(m uint8) {
	result := ((c.A | xaaMagic) & c.X) & m
	c.A = result
	c.setNz(result)
}

func (c *Cpu) lax(m uint8) {
	c.A = m
	c.X = m
	c.setNz(m)
}

// ---- Unstable indexed stores ----

func storeDummyReadAddr(base uint16, index uint8) uint16 {
	return (base & 0xFF00) | ((base + uint16(index)) & 0x00FF)
}

func quirkAddrY(base uint16, y, value uint8) uint16 {
	eff := base + uint16(y)
	if (base & 0xFF00) != (eff & 0xFF00) {
		return (uint16(value) << 8) | (eff & 0x00FF)
	}
	return eff
}

func quirkAddrX(base uint16, x, value uint8) uint16 {
	eff := base + uint16(x)
	if (base & 0xFF00) != (eff & 0xFF00) {
		return (uint16(value) << 8) | (eff & 0x00FF)
	}
	return eff
}

func (c *Cpu) tasStore(b *bus.Bus, base uint16) {
	c.SP = c.A & c.X
	h := uint8(base >> 8)
	value := c.SP & (h + 1)
	storeAddr := quirkAddrY(base, c.Y, value)
	_ = b.Read(storeDummyReadAddr(base, c.Y))
	b.Write(storeAddr, value)
}

func (c *Cpu) ahxStore(b *bus.Bus, base uint16, reg uint8) {
	h := uint8(base >> 8)
	value := reg & (h + 1)
	storeAddr := quirkAddrY(base, c.Y, value)
	_ = b.Read(storeDummyReadAddr(base, c.Y))
	b.Write(storeAddr, value)
}

func (c *Cpu) shyStore(b *bus.Bus, base uint16) {
	h := uint8(base >> 8)
	value := c.Y & (h + 1)
	storeAddr := quirkAddrX(base, c.X, value)
	_ = b.Read(storeDummyReadAddr(base, c.X))
	b.Write(storeAddr, value)
}

// ---- RMW-combo helpers ----

func (c *Cpu) dcp(b *bus.Bus, op Operand) {
	c.rmwCombo(b, op, decValue)
	m := c.readOperand(b, op)
	c.cmpOp(c.A, m)
}

func (c *Cpu) isc(b *bus.Bus, op Operand) {
	c.rmwCombo(b, op, incValue)
	m := c.readOperand(b, op)
	c.sbc(m)
}

func (c *Cpu) slo(b *bus.Bus, op Operand) {
	c.rmwCombo(b, op, aslValue)
	m := c.readOperand(b, op)
	c.oraOp(m)
}

func (c *Cpu) rla(b *bus.Bus, op Operand) {
	c.rmwCombo(b, op, rolValue)
	m := c.readOperand(b, op)
	c.andOp(m)
}

func (c *Cpu) sre(b *bus.Bus, op Operand) {
	c.rmwCombo(b, op, lsrValue)
	m := c.readOperand(b, op)
	c.eorOp(m)
}

func (c *Cpu) rra(b *bus.Bus, op Operand) {
	c.rmwCombo(b, op, rorValue)
	m := c.readOperand(b, op)
	c.adc(m)
}

func (c *Cpu) executeUnofficialRMW(b *bus.Bus, opcode uint8) uint8 {
	switch opcode {
	// ---- DCP ----
	case 0xC7:
		c.dcp(b, c.opZp(b))
		return 5
	case 0xD7:
		c.dcp(b, c.opZpX(b))
		return 6
	case 0xCF:
		c.dcp(b, c.opAbs(b))
		return 6
	case 0xDF:
		c.dcp(b, c.opAbsX(b))
		return 7
	case 0xDB:
		c.dcp(b, c.opAbsY(b))
		return 7
	case 0xC3:
		c.dcp(b, c.opIndX(b))
		return 8
	case 0xD3:
		c.dcp(b, c.opIndY(b))
		return 8

	// ---- ISC ----
	case 0xE7:
		c.isc(b, c.opZp(b))
		return 5
	case 0xF7:
		c.isc(b, c.opZpX(b))
		return 6
	case 0xEF:
		c.isc(b, c.opAbs(b))
		return 6
	case 0xFF:
		c.isc(b, c.opAbsX(b))
		return 7
	case 0xFB:
		c.isc(b, c.opAbsY(b))
		return 7
	case 0xE3:
		c.isc(b, c.opIndX(b))
		return 8
	case 0xF3:
		c.isc(b, c.opIndY(b))
		return 8

	// ---- SLO ----
	case 0x07:
		c.slo(b, c.opZp(b))
		return 5
	case 0x17:
		c.slo(b, c.opZpX(b))
		return 6
	case 0x0F:
		c.slo(b, c.opAbs(b))
		return 6
	case 0x1F:
		c.slo(b, c.opAbsX(b))
		return 7
	case 0x1B:
		c.slo(b, c.opAbsY(b))
		return 7
	case 0x03:
		c.slo(b, c.opIndX(b))
		return 8
	case 0x13:
		c.slo(b, c.opIndY(b))
		return 8

	// ---- RLA ----
	case 0x27:
		c.rla(b, c.opZp(b))
		return 5
	case 0x37:
		c.rla(b, c.opZpX(b))
		return 6
	case 0x2F:
		c.rla(b, c.opAbs(b))
		return 6
	case 0x3F:
		c.rla(b, c.opAbsX(b))
		return 7
	case 0x3B:
		c.rla(b, c.opAbsY(b))
		return 7
	case 0x23:
		c.rla(b, c.opIndX(b))
		return 8
	case 0x33:
		c.rla(b, c.opIndY(b))
		return 8

	// ---- SRE ----
	case 0x47:
		c.sre(b, c.opZp(b))
		return 5
	case 0x57:
		c.sre(b, c.opZpX(b))
		return 6
	case 0x4F:
		c.sre(b, c.opAbs(b))
		return 6
	case 0x5F:
		c.sre(b, c.opAbsX(b))
		return 7
	case 0x5B:
		c.sre(b, c.opAbsY(b))
		return 7
	case 0x43:
		c.sre(b, c.opIndX(b))
		return 8
	case 0x53:
		c.sre(b, c.opIndY(b))
		return 8

	// ---- RRA ----
	case 0x67:
		c.rra(b, c.opZp(b))
		return 5
	case 0x77:
		c.rra(b, c.opZpX(b))
		return 6
	case 0x6F:
		c.rra(b, c.opAbs(b))
		return 6
	case 0x7F:
		c.rra(b, c.opAbsX(b))
		return 7
	case 0x7B:
		c.rra(b, c.opAbsY(b))
		return 7
	case 0x63:
		c.rra(b, c.opIndX(b))
		return 8
	case 0x73:
		c.rra(b, c.opIndY(b))
		return 8

	// ---- TAS / SHS ----
	case 0x9B:
		base := c.fetchWord(b)
		c.tasStore(b, base)
		return 5

	// ---- AHX / SHA ----
	case 0x9F:
		base := c.fetchWord(b)
		c.ahxStore(b, base, c.A&c.X)
		return 5
	case 0x93:
		base := c.amIndirectYBase(b)
		c.ahxStore(b, base, c.A&c.X)
		return 6

	// ---- SHX / SXA ----
	case 0x9E:
		base := c.fetchWord(b)
		c.ahxStore(b, base, c.X)
		return 5

	// ---- SHY / SYA ----
	case 0x9C:
		base := c.fetchWord(b)
		c.shyStore(b, base)
		return 5

	default:
		return 2
	}
}

func (c *Cpu) executeUnofficial(b *bus.Bus, opcode uint8) uint8 {
	switch opcode {
	// ---- Implied NOPs (2 cycles) ----
	case 0x1A, 0x3A, 0x5A, 0x7A, 0xDA, 0xFA:
		return 2

	// ---- Immediate NOPs (2 cycles) ----
	case 0x80, 0x82, 0x89, 0xC2, 0xE2:
		_ = c.fetchByte(b)
		return 2

	// ---- Zero-page NOPs (3 cycles) ----
	case 0x04, 0x44, 0x64:
		_ = c.fetchByte(b)
		return 3

	// ---- Zero-page,X NOPs (4 cycles) ----
	case 0x14, 0x34, 0x54, 0x74, 0xD4, 0xF4:
		_ = c.amZeroPageX(b, DummyRMW)
		return 4

	// ---- Absolute NOP (4 cycles) ----
	case 0x0C:
		_ = c.fetchWord(b)
		return 4

	// ---- Absolute,X NOPs (4 + page-cross) ----
	case 0x1C, 0x3C, 0x5C, 0x7C, 0xDC, 0xFC:
		pc, _ := c.amAbsoluteX(b, DummyRead)
		if pc {
			return 5
		}
		return 4

	// ---- LAX ----
	case 0xA7:
		c.lax(c.rdZp(b))
		return 3
	case 0xB7:
		c.lax(c.rdZpY(b))
		return 4
	case 0xAF:
		c.lax(c.rdAbs(b))
		return 4
	case 0xBF:
		v, pc := c.rdAbsY(b)
		c.lax(v)
		if pc {
			return 5
		}
		return 4
	case 0xA3:
		c.lax(c.rdIndX(b))
		return 6
	case 0xB3:
		v, pc := c.rdIndY(b)
		c.lax(v)
		if pc {
			return 6
		}
		return 5

	// ---- SAX ----
	case 0x87:
		c.wrZp(b, c.A&c.X)
		return 3
	case 0x97:
		c.wrZpY(b, c.A&c.X)
		return 4
	case 0x8F:
		c.wrAbs(b, c.A&c.X)
		return 4
	case 0x83:
		c.wrIndX(b, c.A&c.X)
		return 6

	// ---- ANC ----
	case 0x0B, 0x2B:
		c.anc(c.rdImm(b))
		return 2

	// ---- ALR ----
	case 0x4B:
		c.alr(c.rdImm(b))
		return 2

	// ---- ARR ----
	case 0x6B:
		c.arr(c.rdImm(b))
		return 2

	// ---- AXS / SBX ----
	case 0xCB:
		c.axs(c.rdImm(b))
		return 2

	// ---- XAA (unstable) ----
	case 0x8B:
		c.xaa(c.rdImm(b))
		return 2

	// ---- LAS / LAR ----
	case 0xBB:
		v, pc := c.rdAbsY(b)
		r := v & c.SP
		c.A = r
		c.X = r
		c.SP = r
		c.setNz(r)
		if pc {
			return 5
		}
		return 4

	// ---- KIL / JAM / HLT ----
	case 0x02, 0x12, 0x22, 0x32, 0x42, 0x52, 0x62, 0x72, 0x92, 0xB2, 0xD2, 0xF2:
		c.SetHalted(true)
		return 1

	default:
		return c.executeUnofficialRMW(b, opcode)
	}
}
